javalib/src/main/scala/java/util/ArrayList.scala

```diff
@@ -14,24 +14,51 @@ package java.util
 
 import java.lang.Cloneable
 import java.lang.Utils._
+import java.util.ScalaOps._
 
 import scala.scalajs._
+import scala.scalajs.LinkingInfo.targetPureWasm
 
-class ArrayList[E] private (private[ArrayList] val inner: js.Array[E])
+// _size keeps the track of the effective size of the underlying scala.Array for Wasm
+// in JS, we can use innerJS.length for it's size.
+class ArrayList[E] private (private[ArrayList] var inner: AnyRef, private var _size: Int)
     extends AbstractList[E] with RandomAccess with Cloneable with Serializable {
   self =>
 
+  private val innerJS =
+    if (targetPureWasm) null
+    else inner.asInstanceOf[js.Array[E]]
+
+  // To assign null to elements either when E <: AnyRef or AnyVal,
+  // we use an Array[Any] as an underlying storage and cast to E when needed.
+  private var innerWasm =
```

Comment:
The comment is misleading. It's not really to assign `null` elements. It's because that's the only array type that we can instantiate and that can contain all the required values. In the context of our codebase, and the javalib in particular, it's rather "obvious", though. IMO an explicit type is enough of an explanation:
```suggestion
  private var innerWasm: Array[Any] =
```

Reply:
Ah but to be able to use the methods of `ju.Arrays`, it should be typed as `Array[AnyRef]`. That means we will also need casts when storing to the array, but that's fine.